src/chain/chain_information/babe_fetch_epoch.rs

```diff
@@ -72,27 +70,50 @@ pub fn babe_fetch_epoch(config: Config) -> Query {
     });
 
     match vm {
-        Ok(vm) => Query::from_inner(vm, config.babe_genesis_config),
+        Ok(vm) => Query::from_inner(vm),
         Err(err) => Query::Finished(Err(Error::WasmStart(err))),
     }
 }
 
+/// Partial information about a Babe epoch.
+pub struct PartialBabeEpochInformation {
+    epoch_index: u64,
+    start_slot_number: Option<u64>,
+    authorities: Vec<header::BabeAuthority>,
+    randomness: [u8; 32],
+}
+
+impl PartialBabeEpochInformation {
+    /// Complete the epoch information by supplying Babe configuration information.
+    pub fn complete(
+        self,
+        c: (u64, u64),
+        allowed_slots: header::BabeAllowedSlots,
+    ) -> BabeEpochInformation {
+        BabeEpochInformation {
+            epoch_index: self.epoch_index,
+            start_slot_number: self.start_slot_number,
+            authorities: self.authorities,
+            randomness: self.randomness,
+            c,
+            allowed_slots,
+        }
+    }
+}
```

Comment:
Please do this in `grandpa_warp_sync.rs`.
This module shouldn't touch `BabeEpochInformation` at all. It's objective is purely and solely to do a runtime call.